application/controllers/ApiV1ContactsController.php

```diff
@@ -0,0 +1,604 @@
+                || $matches[1] !== 'application/json'
+            )
+        ) {
+            $this->httpBadRequest('No JSON content');
+        }
+
+        $results = [];
+        $responseCode = 200;
+        $db = Database::get();
+
+        /** @var ?string $identifier */
+        $identifier = $request->getParam('identifier');
+
+        if ($identifier && ! Uuid::isValid($identifier)) {
+            $this->httpBadRequest('The given identifier is not a valid UUID');
+        }
+
+        $filter = FilterProcessor::assembleFilter(
+            QueryString::fromString(rawurldecode(Url::fromRequest()->getQueryString()))
+                ->on(
+                    QueryString::ON_CONDITION,
+                    function (Filter\Condition $condition) {
+                        $column = $condition->getColumn();
+                        if (! in_array($column, ['id', 'full_name', 'username'])) {
+                            $this->httpBadRequest(sprintf(
+                                'Invalid filter column %s given, only id, full_name and username are allowed',
+                                $column
+                            ));
+                        }
+
+                        if ($column === 'id') {
+                            if (! Uuid::isValid($condition->getValue())) {
+                                $this->httpBadRequest('The given filter id is not a valid UUID');
+                            }
+
+                            $condition->setColumn('external_uuid');
+                        }
+                    }
+                )->parse()
+        );
+
+        switch ($method) {
+            case 'GET':
+                $stmt = (new Select())
+                    ->distinct()
+                    ->from('contact co')
+                    ->columns([
+                        'contact_id'        => 'co.id',
+                        'id'                => 'co.external_uuid',
+                        'full_name',
+                        'username',
+                        'default_channel'   => 'ch.name',
+                    ])
+                    ->joinLeft('contact_address ca', 'ca.contact_id = co.id')
+                    ->joinLeft('channel ch', 'ch.id = co.default_channel_id');
+
+                if ($identifier !== null) {
+                    $stmt->where(['external_uuid = ?' => $identifier]);
+
+                    /** @var stdClass|false $result */
+                    $result = $db->fetchOne($stmt);
+
+                    if ($result === false) {
+                        $this->httpNotFound('Contact not found');
+                    }
+
+                    if ($result->username === null) {
+                        unset($result->username);
+                    }
+
+                    $groups = $this->fetchGroupIdentifiers($result->contact_id);
+                    if ($groups) {
+                        $result->groups = $groups;
+                    }
```

Comment:
The omission of the property is a deviation from the behavior of the normal Icinga 2 API.

#### Example response from Icinga 2 when querying a host
```json
{
    "attrs": {
            ...
	    "acknowledgement": 0,
	    "action_url": "",
            ...
	    "check_timeout": null,
            ...
	    "last_check_result": {
                    ...
		    "performance_data": [],
                    ...
	    },
            ...
     },
    "joins": {},
    "meta": {},
    ...
}
```

#### Example response from the `notifications/v1` API when querying a contact without assigned groups
```json
{
    "id": "c63143e2-59a2-4822-b371-d0b039ea2b07",
    "full_name": "Rhys Gibbs",
    "username": "rgibbs",
    "default_channel": "Mail",
    "addresses": "{\"email\":\"rhys.gibbs@local.local\"}"
}
```

This leads to errors when parsing the responses programmatically, as the code would usually expect the field to be existent but empty if there are no assigned groups to the contact in question.
You can also see how Icinga 2 returns empty properties with an empty string `""`, a null value `null` or empty lists `[], {}`.
In my opinion this implementation should behave in the same way as it would keep it consistent to other API endpoints.

Reply:
I have done this on purpose as these fields are optional and should not be present in the response, if they are empty. I have already discussed this with @nilmerg.

Reply:
> I have done this on purpose as these fields are optional and should not be present in the response, if they are empty. I have already discussed this with @nilmerg.

I understand the idea behind your implementation, but it's still inconsistent to Icinga's base API logic.
[It's logic reads the value of the object properties](https://github.com/Icinga/icinga2/blob/bca1a8447af26313123183efa863f947f9590731/lib/remote/objectqueryhandler.cpp#L74-L85) and [provides proper empty values](https://github.com/Icinga/icinga2/blob/master/lib/config/vmops.hpp#L238-L246) like the ones mentioned in my previous comment.

Wouldn't we want to have concistency over the implementations of our own modules?

Reply:
I think our discussion was either about what to expect in requests or not about a specific direction at all. I know that missing keys in such responses are a pain, so I'd also suggest to include empty keys in responses. Requests on the other hand might not need to include such so if there's any optionality involved, it's here.

Reply:
The current PR state now handles the null / missing values correctly.
Example response:
```json
{
    "id": "6d182243-ec20-4b92-be6b-3cc0cbf5671b",
    "full_name": "Rhys Gibbs",
    "username": null,
    "default_channel": "3e73a3dd-c7f2-4245-a8e7-80b2c1cb899d",
    "groups": [],
    "addresses": "{}"
},
```